bumblebee-status: init at 2.2.0

augustebaum opened the pull request:
## Description of changes

This creates a package for `bumblebee-status`, a status program for use with the i3 window manager (or similar).

Because the `bumblebee-status` includes a large number of plugins within its source code (see [here](https://bumblebee-status.readthedocs.io/en/main/modules.html) for the docs and [here](https://github.com/tobi-wan-kenobi/bumblebee-status/tree/main/bumblebee_status/modules) for the source code), each with its own set of dependencies, the package can be overriden in the form of the argument `withPlugins`, which is the list of plugins to get the dependencies for. By default the list is empty (i.e. no extra dependencies are installed, although the program still has access to the plugins).

Hence this PR includes two packages: `bumblebee-status`, and `bumblebee-status-full` for which all plugin dependencies are added as `propagatedBuildInputs`.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

augustebaum commented:
Some things to note:
- Since the plugins are all contained within the bumblebee-status repo, the `withPlugins` override doesn't actually install plugins but rather their `propagatedBuildInputs` (i.e., kind of, runtime dependencies);
- Not all plugins have their dependencies packaged with Nix so I've implemented some logic for marking when a dependency is missing; however, I haven't figured out how to actually print warnings.

andres-nav commented:
First of all, you need to separate the changes in two commits. One adding you as a maintiner, and the other one adding the package, see #233506. Also it is better to get the source code from Github in my opinion. For the tests, it is better to disable the tests that don't work, not all of them. And it should be alphabetically ordered in `all-packages`, but probably as it is a python applicaton it should be in the python directory.

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/plugins.nix:
```diff
@@ -0,0 +1,378 @@
+      inherit name requires requiresUnpackaged optional;
+      warnings =
+        let
+          warn = dep: ''`${dep}` has not been packaged for Nix yet; plugin "${name}" will likely not work as a result.'';
```
Not sure about "packaged for Nix". Maybe you mean "packaged with Nix", or "packaged in Nixpkgs". But I wouldn't say that we package "for" Nix.

Also, stating that this hasn't been done is quite bold. Instead of asserting that it hasn't been done, how about saying that you don't know that it has been done?

```suggestion
          warn = dep: ''`${dep}` is not known to be contained in Nixpkgs. Plugin "${name}" will likely not work as a result.'';
```

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/plugins.nix:
```diff
@@ -0,0 +1,378 @@
+  {
+    name = "brightness";
+    optional = [ brightnessctl light xbacklight ];
+    warnings = [ "If you cannot let this plugin query the system ACPI interface (i.e. the plugin option `use_acpi` is set to `False`, then you need at least one of the optional dependencies." ];
```
* The "I" in "ACPI" stands for "interface", so I suggest to remove "interface".
 * Instead of "let" I suggest "allow" which is a little more explicit and less colloquial.
 * You forgot to close a parenthesis, but I suggest to drop the parenthesis anyway.

```suggestion
    warnings = [ "If you do not allow this plugin to query the system's ACPI, i.e. the plugin option `use_acpi` is set to `False`, then you need at least one of the optional dependencies." ];
```

lorenzleutgeb reviewed (approved):
Very clean job, nice!

One thing I don't understand is why you chose to have a list of plugins. Both the [filtering](https://github.com/NixOS/nixpkgs/pull/254772/files#diff-84a01a5fa30a3465ba9c53b0c37cd7884521371b5c98d8dfea1b421d980c2867R25-R29) and [generation of the list of names of plugins](https://github.com/NixOS/nixpkgs/pull/254772/files#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22R32929) are a little unwieldy. If `plugins.nix` would expose an attribute set, consuming them would be `selectedPlugins = lib.filterAttrs (name: _: builtins.elem name withPlugins) allPlugins`, a list of plugins names would be `allPluginNames = builtins.attrNames allPlugins` and you might even be able to get rid of all the `name = "..."` noise in `plugins.nix`. In case you have a good reason (which apparently I do not comprehend), just ignore this comment.

augustebaum commented:
> One thing I don't understand is why you chose to have a list of plugins

It used to be an attribute set, but I need the `name` for the warning messages... which are currently unused right now. Will think about this some more.

augustebaum committed:
maintainers: add augustebaum

augustebaum committed:
bumblebee-status: init at 2.2.0

augustebaum commented:
Update:
- Added tests (which required switching to `fetchFromGithub`
- Adding myself to the Nixpkgs maintainers list is now a separate commit
- Implemented warnings about non-existent plugin dependencies

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/plugins.nix:
```diff
@@ -0,0 +1,385 @@
+    , requiresUnpackaged ? [ ]
+    , optional ? [ ]
+    , warnings ? [ ]
+    ,
```
```suggestion

```

augustebaum replied:
`nixpkgs-fmt` did that so I'll just leave it as is

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/plugins.nix:
```diff
@@ -0,0 +1,385 @@
+    rec {
+      inherit name requires requiresUnpackaged optional;
+      warnings = warningsList;
+      hasWarnings = warnings != [ ];
+      warningsToString = lib.strings.concatStringsSep " " warnings;
```
Adding these two attributes, which are derived from `warnings` in a simple way, seems strange to me. I'd rather create variables for them wherever you actually *use* them. It turns out the usage actually is in `default.nix`, so you split up the logic for processing warnings into two files. I'd suggest in `plugins.nix` you just define `warnings`, and all further computation should be done in the block for `propagatedBuildInputs` in `default.nix`, using variables as you like.

augustebaum commented:
Hopefully final update: in the end I'd rather not package plugins with missing dependencies at all. The plugins in question are not all relevant for NixOS in any case (e.g. the ones that check the status of some other Linux package manager), but hopefully the remaining ones can be packaged at some point.

imincik reviewed (approved):
Good work @augustebaum

augustebaum commented:
Remove redundant `bumblebee-status-full` package variant.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2697

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/default.nix:
```diff
@@ -0,0 +1,63 @@
+
+  meta = with lib; {
+    description = "bumblebee-status is a modular, theme-able status line generator for the i3 window manager.";
+    homepage = "https://bumblebee-status.readthedocs.io/";
```
```suggestion
    homepage = "https://github.com/tobi-wan-kenobi/bumblebee-status";
```

I'm not sure whether this should refer to the documentation page. I think it is more common to refer to the repository on GitHub in such cases.

augustebaum replied:
Sure, thanks for the pointer!

lorenzleutgeb commented on pkgs/applications/window-managers/i3/bumblebee-status/plugins.nix:
```diff
@@ -0,0 +1,377 @@
+          warn = dep: ''Plugin "${name}": Dependency `${dep}` is not known to be contained in Nixpkgs, so plugin might fail at runtime.'';
+          requiresUnpackagedWarnings = map warn requiresUnpackaged;
+        in
+        warnings ++ requiresUnpackagedWarnings;
```
I am not sure where you're going with this. It seems that you commented all plugins that would trigger a warning? So now this code is somehow dead. Maybe add a comment explaining the situation?

augustebaum replied:
Replaced dead code with a comment.

lorenzleutgeb reviewed (approved):
Only minor comments. Still very clean, and IMO good to merge!

infinisil commented:
We sat together at #ZurichZHF, I have some suggestions against this branch: https://github.com/augustebaum/nixpkgs/pull/1

infinisil reviewed (approved):
Went through some iterations with @augustebaum at #ZurichZHF, looking great now!

augustebaum commented:
Thanks for the thoughtful review!

imincik commented:
Congratulations @augustebaum !
